lib/bats-core/test_functions.bash

```diff
@@ -3,30 +3,89 @@
 BATS_TEST_DIRNAME="${BATS_TEST_FILENAME%/*}"
 BATS_TEST_NAMES=()
 
-# Shorthand for source-ing files relative to the BATS_TEST_DIRNAME,
-# optionally with a .bash suffix appended. If the argument doesn't
-# resolve relative to BATS_TEST_DIRNAME it is sourced as-is.
+# _load expects an absolute path to a file or directory to source.
+# If the argument is a file the file will be sourced or the same path
+# with the added suffix `.bash` exists it will be sourced.
+# If the argument is a directory _load checks of the file `load` or
+# `load.bash` inside - if it exists this file is loaded. If it doesn't
+# exist all files with the `.bash` suffix are sourced.
+#
+# Order with `/path/to/example` as argument:
+#  - /path/to/example.bash
+#  - /path/to/example
+#  - /path/to/example/load.bash
+#  - /path/to/example/load
+#  - /path/to/example/*.bash
+_load() {
+    local file="${1:?}"
+
+    if [[ "${file:0:1}" != "/" ]]; then
+        printf "Received argument with a relative path, expected absolute: %s\n" "$file" >&2
+        return 1
+    fi
+
+    local -a opts=(
+        "$file.bash"
+        "$file"
+        "$file/load.bash"
+        "$file/load"
+    )
+
+    for opt in "${opts[@]}"; do
+        if [[ -f "$opt" ]]; then
+            if ! source "$opt"; then
+                printf 'Sourcing file "%s" failed' "$contained" >&2
+                exit 1
+            fi
+            return
+        fi
+    done
+
+    if [[ -d "$file" ]]; then
+        for contained in "$file"/*.bash; do
+            if ! source "$contained"; then
+                printf 'Sourcing file "%s" failed' "$contained" >&2
+                exit 1
+            fi
+        done
+        return
+    fi
+
+    return 1
+}
+
+# Shorthand to source files relative to the test file
+# (BATS_TEST_DIRNAME) and from BATS_LIB_PATH.
 load() {
   local file="${1:?}"
 
-  # For backwards-compatibility first look for a .bash-suffixed file.
-  # TODO consider flipping the order here; it would be more consistent
-  # and less surprising to look for an exact-match first.
-  if [[ -f "${BATS_TEST_DIRNAME}/${file}.bash" ]]; then
-    file="${BATS_TEST_DIRNAME}/${file}.bash"
-  elif [[ -f "${BATS_TEST_DIRNAME}/${file}" ]]; then
-    file="${BATS_TEST_DIRNAME}/${file}"
+  # Check if target is absolute
+  if [[ "${file:0:1}" == "/" ]]; then
+    if ! _load "$file"; then
+      printf "Failed to load file or library '%s'\n" "$file" >&2
+      return 1
+    fi
+    return
```

Comment:
Why should this be a fatal error in _load?
If _load returns a non-zero returncode (or exits) the conditional will print a message and return.
_load returns a non-zero returncode when it can't find the targeted file/library.
Otherwise the file at the absolute path has been sourced correctly and load can return.

Reply:
IIRC in a previous patchset there was an `exit 1` around this line https://github.com/bats-core/bats-core/pull/27/files#diff-223055a1b11f29286d5c612d214344c7R159

Apologies if that was a mis-remember. Current patchset looks correct WRT return codes, will test again.